File: ex2.py
from collections import defaultdict

# 定義一個二分圖匹配算法
def bpm(graph, u, matchR, seen):
    # 嘗試為 u 配對
    for v in graph[u]:
        if not seen[v]:
            seen[v] = True
            # 如果 v 沒有被配對，或被配對的 u 可以換其他匹配
            if matchR[v] == -1 or bpm(graph, matchR[v], matchR, seen):
                matchR[v] = u
                return True
    return False

def maxBPM(graph, N1, N2):
    # matchR[v] 是與 v 配對的 u，初始值為 -1
    matchR = [-1] * N2
    result = 0
    for u in range(N1):
        # 為每個 u 嘗試配對
        seen = [False] * N2
        if bpm(graph, u, matchR, seen):
            result += 1
    return result

# 讀取輸入
N1, N2, M = map(int, input().split())
graph = defaultdict(list)

# 讀取不能一起服用的組合
for _ in range(M):
    i, j = map(int, input().split())
    graph[i - 1].append(j - 1)  # 將索引轉成從 0 開始

# 計算最大匹配
max_pairs = maxBPM(graph, N1, N2)

# 輸出結果
print(N1 + N2 - max_pairs)

File: ex2.cpp
#include <iostream>
#include <vector>
#include <cstring>

using namespace std;

// 定義最大數量
const int MAXN = 100;

vector<int> graph[MAXN];  // 存儲圖
int match[MAXN];  // 配對結果
bool visited[MAXN];

// 匈牙利算法的 DFS，嘗試為 u 找到匹配
bool bpm(int u) {
    for (int v : graph[u]) {
        if (!visited[v]) {
            visited[v] = true;
            // 如果 v 還沒有匹配，或是 v 原來的匹配對象可以找到其他匹配
            if (match[v] == -1 || bpm(match[v])) {
                match[v] = u;
                return true;
            }
        }
    }
    return false;
}

// 主函數：計算最大匹配
int maxBPM(int N1, int N2) {
    // 初始化匹配結果
    memset(match, -1, sizeof(match));
    int result = 0;
    
    // 嘗試為每一個眼睛保健營養品找到匹配
    for (int u = 0; u < N1; u++) {
        memset(visited, false, sizeof(visited));
        if (bpm(u)) {
            result++;
        }
    }
    
    return result;
}

int main() {
    int N1, N2, M;
    cin >> N1 >> N2 >> M;  // 讀取 N1, N2, M
    
    // 讀取不能同時服用的組合
    for (int i = 0; i < M; i++) {
        int x, y;
        cin >> x >> y;
        x--;  // 調整為從 0 開始的索引
        y--;
        graph[x].push_back(y);  // 建立 x 到 y 的邊
    }

    // 計算最大匹配
    int max_pairs = maxBPM(N1, N2);
    
    // 輸出結果
    cout << N1 + N2 - max_pairs << endl;
    
    return 0;
}
